cmd/dp-filter-api/main.go

```diff
@@ -101,79 +113,125 @@ func main() {
+			logIfError(ctx, observationStore.Close(ctx), "unable to close observation store")
 		}
 
 		if serviceList.FilterOutputSubmittedProducer {
-			log.Info("closing filter output submitted producer", nil)
+			log.Event(ctx, "closing filter output submitted producer")
 			// Close producer after http server has closed so if a message
 			// needs to be sent to kafka off a request it can
-			logIfError(producer.Close(ctx), "unable to close filter output submitted producer")
+			logIfError(ctx, producer.Close(ctx), "unable to close filter output submitted producer")
 		}
 
 		if serviceList.AuditProducer {
-			log.Info("closing audit producer", nil)
-			logIfError(auditProducer.Close(ctx), "unable to close audit producer")
+			log.Event(ctx, "closing audit producer")
+			logIfError(ctx, auditProducer.Close(ctx), "unable to close audit producer")
 		}
 	}()
 
 	// wait for shutdown success (via cancel) or failure (timeout)
 	<-ctx.Done()
 
-	log.Info("Shutdown complete", nil)
+	log.Event(ctx, "Shutdown complete")
 	os.Exit(1)
 }
 
-func exitIfError(err error, message string) {
+func startHealthCheck(ctx context.Context, cfg *config.Config, datasetAPI *dataset.Client, producer *kafka.Producer, observationStore *graph.DB, dataStore *mongo.FilterStore, auditProducer *kafka.Producer) healthcheck.HealthCheck {
+
+	hasErrors := false
+
+	versionInfo, err := healthcheck.NewVersionInfo(BuildTime, GitCommit, Version)
+	if err != nil {
+		log.Event(ctx, "error creating version info", log.FATAL, log.Error(err))
+		hasErrors = true
+	}
+
+	hc := healthcheck.New(versionInfo, cfg.HealthCheckCriticalTimeout, cfg.HealthCheckInterval)
+
+	if err = hc.AddCheck("Dataset API", datasetAPI.Checker); err != nil {
+		log.Event(ctx, "error creating dataset API health check", log.Error(err))
+		hasErrors = true
+	}
+
+	if err = hc.AddCheck("Kafka Producer", producer.Checker); err != nil {
+		log.Event(ctx, "error adding check for kafka producer", log.ERROR, log.Error(err))
+		hasErrors = true
+	}
+
+	if err = hc.AddCheck("GraphDB", observationStore.Checker); err != nil {
+		hasErrors = true
+		log.Event(ctx, "error creating graph db connection", log.ERROR, log.Error(err))
+	}
+
+	checkMongoClient := dataStore.HealthCheckClient()
+	if err = hc.AddCheck("MongoDB", checkMongoClient.Checker); err != nil {
+		log.Event(ctx, "error creating mongodb health check", log.ERROR, log.Error(err))
+		hasErrors = true
+	}
+
+	if cfg.EnablePrivateEndpoints {
+		if err = hc.AddCheck("Kafka Audit Producer", auditProducer.Checker); err != nil {
+			log.Event(ctx, "error adding check for kafka audit producer", log.ERROR, log.Error(err))
+			hasErrors = true
+		}
+
+		// zebedee is used only for identity checking
+		zebedeeClient := zebedee.New(cfg.ZebedeeURL)
+		if err = hc.AddCheck("Zebedee", zebedeeClient.Checker); err != nil {
+			log.Event(ctx, "error creating zebedee health check", log.ERROR, log.Error(err))
+			hasErrors = true
+		}
+	}
+
+	if hasErrors {
+		os.Exit(1)
```

Comment:
maybe return error and handle this in call to this function

Reply:
I considered that, though I thought the code would be cleaner like this, as we would have to create a new error instance to return and we can maintain just one return value?

Reply:
Yes I think the direction from team is that main should handle all errors and be doing a lot less, e.g. use run() function to setup api server and graceful shutdown but return errors from run back to main so there is one exit point of the application. If we are heading down that route then we would want to return an error here an exit app back in main function

I will leave this up to you for now, if we do end up using `run(...) error {` then the team will need to update all applications anyway